### PR TITLE
core/cmds/io: Add cmos commands to io cmd

### PR DESCRIPTION
**Pull request by andrewsun2898**
An alternative to the last PR. Since the implementation of cmos write/read is similar to the io commands, it might be cleaner to just extend the io command. Doing this allows us to leverage the io command's chaining feature as well without needing to reimplement it.

### CODE REVIEW

File: pkg/cmos/cmos.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package cmos
+
+import (
+	"errors"
+	"github.com/u-root/u-root/pkg/memio"
+)
+
+func Read(reg int64, data memio.UintN) error {
+	regVal := memio.Uint8(reg)
```
Comment (rminnich) [inline pkg/cmos/cmos.go:15]: Also, use fmt.Errorf and, also, a message like "byte %d is inside the range 0-13 which is reserved for RTC" is certainly a bit easier to debug than ""bytes below 14 are reserved for RTC"

File: pkg/cmos/cmos.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	if regVal%128 < 14 {
+		return errors.New("bytes below 14 are reserved for RTC")
+	}
+	if err := memio.Out(0x70, &regVal); err != nil {
```
Comment (rminnich) [inline pkg/cmos/cmos.go:19]: maybe add a named constant to the memio package? 
CMOSAddress
CMOSValue?

File: pkg/cmos/cmos.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+package cmos
+
+import (
+	"fmt"
+	"github.com/u-root/u-root/pkg/memio"
+)
+
+const (
+	CMOS_REG_PORT  = 0x70
```
Comment (hugelgupf) [inline pkg/cmos/cmos.go:15]: Go doesn't usually like constants like these. We tolerate them where they reflect some other code base and are really commonly used (Linux kernel constants, etc). I think in this case I'd prefer camelcase https://golang.org/doc/effective_go.html#mixed-caps

File: pkg/cmos/cmos.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2012-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package cmos lets you read and write to cmos registers while doing basic checks on valid register selections.
+
```
Comment (hugelgupf) [inline pkg/cmos/cmos.go:6]: no empty line between the package comment and package statement. stmt must immediately follow comment, or the godoc parser won't pick it up

File: pkg/cmos/cmos.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2012-2019 the u-root Authors. All rights reserved
```
Comment (hugelgupf) [inline pkg/cmos/cmos.go:1]: 2020

File: pkg/cmos/cmos.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package cmos lets you read and write to cmos registers while doing basic checks on valid register selections.
+
+package cmos
+
+import (
+	"fmt"
+	"github.com/u-root/u-root/pkg/memio"
```
Comment (hugelgupf) [inline pkg/cmos/cmos.go:11]: usually, stdlib imports are separated from other imports

```go
import (
  "fmt"

  "github.com/u-r..."
)
```

File: pkg/cmos/cmos.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+const (
+	CMOS_REG_PORT  = 0x70
+	CMOS_DATA_PORT = 0x71
+)
+
+func Read(reg int64, data memio.UintN) error {
```
Comment (hugelgupf) [inline pkg/cmos/cmos.go:19]: Leave a comment on every exported symbol that starts with the symbol name. Should consist of full sentences: https://golang.org/doc/effective_go.html#commentary

These get parsed and put into godocs. (No, our code base is terrible at doing this consistently; we're trying to apply a consistent standard to new code reviews, though. This is equivalent to internal Go readability.)

```go
// Read reads a register reg from CMOS into data.
func Read...
```

File: cmds/core/io/cmos.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// +build amd64 386
+
+package main
+
+import (
+	"github.com/u-root/u-root/pkg/cmos"
+)
+
+func init() {
+	usageMsg += `io (cr index)...
```
Comment (hugelgupf) [inline cmds/core/io/cmos.go:14]: can we add what they mean at the end of the line?

e.g.
```bash
$ ./io
io (r{b,w,l,q} address)...
io (w{b,w,l,q} address value)...
io (in{b,w,l} address)...
io (out{b,w,l} address value)...
io (cr index)...
io (cw index value)...
```

to
```bash
$ ./io
...
io (cr index)... # read from CMOS register index
io (cw index value)... # write value to CMOS register index
```

File: cmds/core/io/cmos.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2012-2019 the u-root Authors. All rights reserved
```
Comment (hugelgupf) [inline cmds/core/io/cmos.go:1]: 2020

File: pkg/cmos/cmos.go (hunk 3 above)
Comment (insomniacslk) [reply pkg/cmos/cmos.go]: I guess this is my fault from the other PR

File: pkg/cmos/cmos.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+		return err
+	}
+	return memio.In(cmosDataPort, data)
+}
+
+// Write writes value data into CMOS register reg.
+func Write(reg int64, data memio.UintN) error {
+	regVal := memio.Uint8(reg)
+	if regVal%128 < 14 {
+		return fmt.Errorf("byte %d is inside the range 0-13 which is reserved for RTC", regVal)
```
Comment (hugelgupf) [inline pkg/cmos/cmos.go:35]: Why are we not allowing RTC?

Comment (andrewsun2898) [reply pkg/cmos/cmos.go:35]: Our internal tool for modifying the cmos doesn't allow it. I'm not too sure what the impacts of changing the rtc bytes are so if nobody has any issues with it I can remove the check.

Comment (andrewsun2898) [reply pkg/cmos/cmos.go:35]: Alternatively, I could add another function to cmds/core/io/cmos.go for read and write and perform the checks there. That way the pkg functions can be more flexible for future references while preserving the checks for the command.

Comment (rminnich) [reply pkg/cmos/cmos.go:35]: or you could add a rtcr and rtcw command which explicitly writes the rtc, i.e. only allows the range 0 to 13?

**Comment by hugelgupf** [comment]
I think maybe it makes sense to single out RTC for a command-line tool, but not a reusable Go package.

Even then, the io tool doesn't place restrictions on any IO ports, I wouldn't expect any for CMOS either.

**Comment by hugelgupf** [comment]
To rephrase my objection - it's leaky abstraction. The io tool and the lowest level Read/Write primitive don't usually concern themselves with the content of something. They write what you ask them to write, unconditionally.

If you want something that interprets the meaning of those registers, I'd write another layer/set of tools in the package, ReadRTC/WriteRTC. 

If you wanna commit this as is, that's fine, too, Andrew - the decision is yours.

File: cmds/core/io/cmos.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+`
+	readCmds["cr"] = cmd{cmosRead, 7, 8}
+	readCmds["rtcr"] = cmd{rtcRead, 7, 8}
+	writeCmds["cw"] = cmd{cmosWrite, 7, 8}
+	writeCmds["rtcw"] = cmd{rtcWrite, 7, 8}
+}
+
+func cmosRead(reg int64, data memio.UintN) error {
+	regVal := memio.Uint8(reg)
+	if regVal%128 < 14 {
```
Comment (rminnich) [inline cmds/core/io/cmos.go:30]: why the mod? I don't see the need.

Comment (andrewsun2898) [reply cmds/core/io/cmos.go:30]: woops yep your right.

Comment (rminnich) [reply cmds/core/io/cmos.go]: the address is guaranteed to be 7 bits ... you said that on line 22.

Comment (andrewsun2898) [reply cmds/core/io/cmos.go]: Yep, copy & pasted without thinking, but you're right

File: cmds/core/io/cmos.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+	readCmds["cr"] = cmd{cmosRead, 7, 8}
+	readCmds["rtcr"] = cmd{rtcRead, 7, 8}
+	writeCmds["cw"] = cmd{cmosWrite, 7, 8}
+	writeCmds["rtcw"] = cmd{rtcWrite, 7, 8}
+}
+
+func cmosRead(reg int64, data memio.UintN) error {
+	regVal := memio.Uint8(reg)
+	if regVal < 14 {
+		return fmt.Errorf("byte %d is inside the range 0-13 which is reserved for RTC", regVal)
```
Comment (rminnich) [inline cmds/core/io/cmos.go:31]: yay! an error message! Linux just gives you EINVAL on stuff like this, it's nice to have a meaningful error :-)

File: pkg/cmos/cmos.go (hunk 4 above)
Comment (hugelgupf) [reply pkg/cmos/cmos.go:5]: argh, sorry, i meant the other way.

```go
// copyright ...

// Package cmos...
package cmos
```

**Pull request merged**